pkg/transport/pool.go

```diff
@@ -58,17 +100,34 @@ func (tp *TransportPool) newClient(name string) (*Client, error) {
 	}
 
 	sub.SetDefault(paramTransportClientTimeout, defaultTransportClientTimeout)
+	sub.SetDefault(paramTransportCompress, defaultTransportCompress)
+	sub.SetDefault(paramTransportCustomHeaders, nil) // No good way to express this as a const that I can find.
+	sub.SetDefault(paramTransportDebugBody, defaultTransportDebugBody)
+	sub.SetDefault(paramTransportEnableMetrics, defaultTransportEnableMetrics)
+	sub.SetDefault(paramTransportMaxParallelRequests, defaultTransportMaxParallelRequests)
 	sub.SetDefault(paramTransportType, defaultTransportType)
+	sub.SetDefault(paramTransportUserAgent, defaultTransportUserAgent)
 
 	clientTimeout := sub.GetDuration(paramTransportClientTimeout)
-	transportType := sub.Get(paramTransportType)
+	compress := sub.GetBool(paramTransportCompress)
+	customHeaders := sub.GetStringMapString(paramTransportCustomHeaders)
+	debugBody := sub.GetBool(paramTransportDebugBody)
+	enableMetrics := sub.GetBool(paramTransportEnableMetrics)
+	maxParallelRequests := sub.GetInt(paramTransportMaxParallelRequests)
+	transportType := sub.GetString(paramTransportType)
+	userAgent := sub.GetString(paramTransportUserAgent)
 
 	if clientTimeout < 0 {
 		return nil, errors.New(paramTransportClientTimeout + " must not be negative") // 0 = no timeout
 	}
 
+	backoff, err := util.GetRetryFromViper(sub)
+	if err != nil {
+		tp.logger.WithField("name", name).WithError(err).Error("failed to load retry policy")
+		return nil, err
```

Comment:
Is there a reason to both log and return the error? Ideally should only return it

Reply:
Not specifically.  Depending on the call stack it could actually end up dropping a few duplicate logs - they're all fatal though (fail early), so it's something I didn't put a lot of thought in to.

For preference I'd actually make it a `.Fatal()`. I've been slowly pushing to remove errors from the code (eg, this whole refactor is fire and forget) with a view to either fail-fast on startup, or handle runtime errors immediately when there's maximum context available.

Ultimately a lot of the error paths just pop out somewhere high level with a log line, and they lose context that could be supplied by an immediate log (such as the transport name in this case).

There's an argument to be made for more strongly typed errors to provide context, but a) I'm lazy, and b) the only way to expose that context is through `error.Error()` which means it's not exposed as structured log data.

Reply:
Ok, I see your point. I think this code can be improved further if config parsing is moved into `NewTransportPool()` and viper is no longed referenced/used. I don't know if it's easy to do this. Also, just terminating the program in the middle of the pooling code would be weird. Testing and reusing this code becomes harder.